support `GrpAbFinGen` groups in group functions

ThomasBreuer opened the pull request:
- added "group function" methods for `GrpAbFinGen` groups (I am not sure how much of this should better be moved to Hecke)

- changed `GroupConjClass` to be an abstract type, defined `GAPGroupConjClass` and `GrpAbFinGenConjClass`

- renamed `radical_subgroup` to `solvable_radical` (the name has meanwhile been corrected in GAP)

- support prescribed type of integer for the return value of `is_pgroup_with_prime`, `length(::GroupConjClass)`, `number_conjugacy_classes`

- fixed `hall_subgroup_reps`

- added `images` to compute the image of a subgroup under a homomorphism

- added some doctests, improved docstrings

ThomasBreuer committed:
support `GrpAbFinGen` groups in group functions

- added "group function" methods for `GrpAbFinGen` groups
  (I am not sure how much of this should better be moved to Hecke)

- changed `GroupConjClass` to be an abstract type,
  defined `GAPGroupConjClass` and `GrpAbFinGenConjClass`

- renamed `radical_subgroup` to `solvable_radical`
  (the name has meanwhile been corrected in GAP)

- support prescribed type of integer for the return value of
  `is_pgroup_with_prime`, `length(::GroupConjClass)`,
  `number_conjugacy_classes`

- fixed `hall_subgroup_reps`

- added `images` to compute the image of a subgroup under a homomorphism

- added some doctests, improved docstrings

ThomasBreuer committed:
adjust further

thofma commented on src/Groups/GAPGroups.jl:
```diff
@@ -994,13 +1046,17 @@ the intersection of all maximal subgroups of `G`.
+
+@deprecate radical_subgroup solvable_radical
+@deprecate has_radical_subgroup has_solvable_radical
+@deprecate set_radical_subgroup set_solvable_radical
```
I think there is a file where we collect them?

fingolfin replied:
Indeed, `src/Deprecations.jl`

We also have some in `src/Groups/sub.jl` (I think they may predate `src/Deprecations.jl`)

thofma commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+   if is_subgroup(H, K)[1] && is_subgroup(K, H)[1]
+     return true, zero(G)
+   else
+     return false, nothing
```
```suggestion
     return false, zero(G)
```
is there harm in doing that?

ThomasBreuer replied:
Currently the `representative_action` documentation promises `nothing`, and the method for `GAPGroup` does this.
We can change the definition and the methods such that always the identity element of the group is returned instead of `nothing`.

thofma commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+#
+################################################################################
+
+#TODO: how to compute complements?
```
Complements are already in `Hecke`. Probably `has_complex(U, G)` or something like that.

ThomasBreuer replied:
There is `has_complement`, which computes one complement if it exists. `complement_class_reps` would have to compute all complements.

thofma replied:
@fieker I remember you played with complements. Do you by accident know how to "efficiently" compute all representatives?

thofma commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+end
+
+function prime_of_pgroup(::Type{T}, G::GrpAbFinGen) where T <: IntegerUnion
+  return ZZRingElem(_prime_of_pgroup(G))
```
```suggestion
  return T(_prime_of_pgroup(G))
```

thofma reviewed (commented):
Looks good! I left a few remarks.

fingolfin commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+# complement_class_reps(G::T, N::T) where T <: GrpAbFinGen
+# complement_system(G::GrpAbFinGen)
+
+function sylow_subgroup(G::GrpAbFinGen, p::IntegerUnion)
```
Note that there also is `function psylow_subgroup(G::GrpAbFinGen, p::IntegerUnion, to_lattice::Bool = true)`

thofma replied:
Yes, we should consolidate this

ThomasBreuer replied:
What about supporting the name `sylow_subgroup` in Hecke?
(And I think the `to_lattice` hint should be added to all functions that call `sub` for their return values.)

thofma replied:
`sylow_subgroup` sounds better. I will add it in Hecke.

fingolfin commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+fitting_subgroup(G::GrpAbFinGen) = (G, identity_map(G))
+
+function frattini_subgroup(G::GrpAbFinGen)
+   is_finite(G) || throw(ArgumentError("G is not finite"))
```
Could also use the `@req` macro:
```suggestion
   @req is_finite(G) "G is not finite"
```

fingolfin commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+
+has_gens(G::GrpAbFinGen) = true
+
+#TODO: improve?
```
Maybe `snf` can be used, even to get a minimal generating set, see http://www.thofma.com/Hecke.jl/stable/abelian/structural/#snf-Tuple{GrpAbFinGen}

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -102,7 +102,28 @@ positive integer `n` such that `x^n` is the identity of `G`.
+
+julia> g = free_group(1);
+
+julia> isfinite(g)
```
Meta-question: should we say `is_finite` here, to be more consistent with the rest? (I have no strong opinion on this, just noticed the contrast to the next input line)

fingolfin commented on src/Groups/GAPGroups.jl:
```diff
@@ -1098,8 +1154,7 @@ function hall_subgroup_reps(G::GAPGroup, P::AbstractVector{<:IntegerUnion})
 
-hall_subgroups_representatives(G::GAPGroup, P::AbstractVector{<:IntegerUnion}) = hall_subgroup_reps(G,P)
-# use @alias?
+@alias hall_subgroups_representatives hall_subgroup_reps
```
Should this be in `src/Aliases.jl` perhaps?

fingolfin commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+
+number_conjugacy_classes(G::GrpAbFinGen) = ZZRingElem(order(G))
+
+number_conjugacy_classes(::Type{T}, G::GrpAbFinGen) where T <: IntegerUnion = T(order(G))
```
```suggestion
number_conjugacy_classes(::Type{T}, G::GrpAbFinGen) where T <: IntegerUnion = order(T, G)
```

fingolfin commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+
+Base.rand(rng::Random.AbstractRNG, C::GrpAbFinGenConjClass) = representative(C)
+
+number_conjugacy_classes(G::GrpAbFinGen) = ZZRingElem(order(G))
```
```suggestion
number_conjugacy_classes(G::GrpAbFinGen) =order(ZZRingElem, G)
```

fingolfin commented on src/Groups/GrpAb.jl:
```diff
@@ -12,3 +12,348 @@ function restrict_codomain(f::GrpAbFinGenMap)
+Base.:^(G::GrpAbFinGen, x::GrpAbFinGenElem) = G
+
+function is_conjugate(G::GrpAbFinGen, H::GrpAbFinGen, K::GrpAbFinGen)
+   if is_subgroup(H, K)[1] && is_subgroup(K, H)[1]
```
Should this also have a test whether `H` and `K` are actually subgroups of `G`?

Out of curiosity, does `H == K` also work? With all those magic embeddings, this is not clear to me.

ThomasBreuer replied:
No, `H == K` is not implied by `is_subgroup(H, K)[1] && is_subgroup(K, H)[1]`.

fingolfin replied:
Yeah I guessed as much. Very confusing :-(

ThomasBreuer replied:
For group theoretic questions, it is usual to construct subgroups in different ways and then ask whether they are equal. From this point of view, the setup for `GrpAbFinGen` is indeed strange.
Here is an example.
```
julia> g = abelian_group([2,3]);  # cyclic group of order 6

julia> s1, emb1 = sylow_subgroup(g, 3);  # the unique subgroup of order 3

julia> s2 = low_index_subgroup_reps(g, 2)[2]; # another way to get the subgroup of order 3

julia> flag2, emb2 = is_subgroup(s2, g); flag2
true

julia> order(s1) == order(s2)
true

julia> is_subgroup(s1, s2)[1] && is_subgroup(s2, s1)[1]  # we can check containment ("magically", w.r.t. the embeddings into g)
true

julia> s1 == s2  # the two groups are not regarded as equal
false
```
The point seems to be that `s1` and `s2` are regarded as independent *groups*, and that only `emb1` and `emb2` interpret them as *subgroups* of `g`. Perhaps the cleanest way to ask for the equality as subgroups of `g` would be something like `is_equal_image(emb1, emb2)`.

ThomasBreuer committed:
address the comments

- moved `@alias` and `@deprecate` to the files where they are collected

- use `@req` where possible

- added missing functions for which Hecke provides an implementation

- write `is_finite` not `isfinite` (thus the `@alias` for `has_isfinite` gets reversed)

- fixed wrong signature for `low_index_subgroup_reps(G::GAPGroup, n::Int)`

ThomasBreuer committed:
three more improvements according to the comments

ThomasBreuer committed:
added a missing `order` method

ThomasBreuer committed:
adjusted the error messages in docstrings

ThomasBreuer committed:
write `is_finite` not `Base.isfinite`

(O.k., I agree that it is nicer now.
But why does Documenter think that it is an error?
I get `Base.isfinite === is_finite`.)

ThomasBreuer commented:
@thofma @fingolfin How shall we proceed with this pull request?
Eventually part of  it should be moved to Hecke, but are there reasons not to merge it into Oscar?

thofma commented:
Looks good to me. @fingolfin can merge if he is happy

fingolfin reviewed (approved):
Thank you!

fingolfin commented:
Can we close #161 now? Maybe @ThomasBreuer can have a look, and either close it, or perhaps summarize in a comment or a new issue what is still missing?
